### ISSUE TITLE
[Bug]  ValueError: Parameter model.layers.30.mlp.down_proj.q_weight has shape (5120, 1712), but expected (5120, 1728)

### ISSUE THREAD

**Issue by Tao-begd** [mlc-ai/mlc-llm#1456]
## 🐛 Bug

I got this error while compiling my retrained llama2-13b model.
My model was converted using the [convert_llama_weights_to_hf.py ](https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/convert_llama_weights_to_hf.py)method from huggingface.
Now I want to ask, is there any way to change the shape of the tensor or any other way to avoid this error.

## To Reproduce

1.mlc_chat  convert_weight --model ./dist/models/Llama2-13b-chat-hf/ --quantization q4f16_1 --model-type llama -o dist/OUTPUT/params

## Expected behavior

> [2023-12-18 16:44:46] INFO auto_config.py:62: Found model configuration: dist\models\Llama2-13b-chat-hf\config.json
> [2023-12-18 16:44:46] INFO auto_device.py:37: Using device: vulkan:0. Use `--device` to override.
> [2023-12-18 16:44:46] INFO auto_weight.py:70: Finding weights in: dist\models\Llama2-13b-chat-hf
> [2023-12-18 16:44:46] INFO auto_weight.py:120: Found source weight format: huggingface-torch. Source configuration: dist\models\Llama2-13b-chat-hf\pytorch_model.bin.index.json
> [2023-12-18 16:44:46] INFO auto_weight.py:139: Not found Huggingface Safetensor
> [2023-12-18 16:44:46] INFO auto_weight.py:106: Using source weight configuration: dist\models\Llama2-13b-chat-hf\pytorch_model.bin.index.json. Use `--source` to override.
> [2023-12-18 16:44:46] INFO auto_weight.py:110: Using source weight format: huggingface-torch. Use `--source-format` to override.
> Weight conversion with arguments:
>   --config          dist\models\Llama2-13b-chat-hf\config.json
>   --quantization    GroupQuantize(name='q4f16_1', kind='group-quant', group_size=32, quantize_dtype='int4', storage_dtype='uint32', model_dtype='float16', num_elem_per_storage=8, num_storage_per_group=4, max_int_value=7)
>   --model-type      llama
>   --device          vulkan:0
>   --source          dist\models\Llama2-13b-chat-hf\pytorch_model.bin.index.json
>   --source-format   huggingface-torch
>   --output          dist\OUTPUT\params
> [2023-12-18 16:44:46] INFO llama_model.py:42: context_window_size not found in config.json. Falling back to max_position_embeddings (2048)
> [2023-12-18 16:44:49] INFO huggingface_loader.py:169: Loading HF parameters from: dist\models\Llama2-13b-chat-hf\pytorch_model-00003-of-00003.bin
> [2023-12-18 16:45:58] INFO group_quantization.py:204: Compiling quantize function for key: (64000, 5120, 'float16', 'vulkan')
> [2023-12-18 16:46:03] INFO huggingface_loader.py:121: [Quantized] Parameter: "lm_head.q_weight", shape: (64000, 640), dtype: uint32
> [2023-12-18 16:46:03] INFO huggingface_loader.py:121: [Quantized] Parameter: "lm_head.q_scale", shape: (64000, 160), dtype: float16
> [2023-12-18 16:46:03] INFO huggingface_loader.py:129: [Not quantized] Parameter: "model.layers.30.input_layernorm.weight", shape: (5120,), dtype: float16
> [2023-12-18 16:46:04] INFO group_quantization.py:204: Compiling quantize function for key: (5120, 13696, 'float16', 'vulkan')
> [2023-12-18 16:46:04] INFO huggingface_loader.py:121: [Quantized] Parameter: "model.layers.30.mlp.down_proj.q_weight", shape: (5120, 1712), dtype: uint32
>   1%|▋                                                                               | 2/243 [01:14<2:30:03, 37.36s/it]
> Traceback (most recent call last):
>   File "C:\Users\admin\AppData\Local\conda\conda\envs\mlc-llm\lib\runpy.py", line 196, in _run_module_as_main
>     return _run_code(code, main_globals, None,
>   File "C:\Users\admin\AppData\Local\conda\conda\envs\mlc-llm\lib\runpy.py", line 86, in _run_code
>     exec(code, run_globals)
>   File "C:\Users\admin\AppData\Local\conda\conda\envs\mlc-llm\Scripts\mlc_chat.exe\__main__.py", line 7, in <module>
>   File "C:\Users\admin\AppData\Local\conda\conda\envs\mlc-llm\lib\site-packages\mlc_chat\__main__.py", line 33, in main
>     cli.main(sys.argv[2:])
>   File "C:\Users\admin\AppData\Local\conda\conda\envs\mlc-llm\lib\site-packages\mlc_chat\cli\convert_weight.py", line 86, in main
>     convert_weight(
>   File "C:\Users\admin\AppData\Local\conda\conda\envs\mlc-llm\lib\site-packages\mlc_chat\compiler\convert_weight.py", line 147, in convert_weight
>     _convert_args(args)
>   File "C:\Users\admin\AppData\Local\conda\conda\envs\mlc-llm\lib\site-packages\mlc_chat\compiler\convert_weight.py", line 101, in _convert_args
>     _check_param(name, param)
>   File "C:\Users\admin\AppData\Local\conda\conda\envs\mlc-llm\lib\site-packages\mlc_chat\compiler\convert_weight.py", line 82, in _check_param
>     raise ValueError(
> ValueError: Parameter model.layers.30.mlp.down_proj.q_weight has shape (5120, 1712), but expected (5120, 1728)

## Environment

 - Platform (e.g. WebGPU/Vulkan/IOS/Android/CUDA): Vulkan
 - Operating system (e.g. Ubuntu/Windows/MacOS/...): Windows
 - Device (e.g. iPhone 12 Pro, PC+RTX 3090, ...)
 - How you installed MLC-LLM (`conda`, source): conda
 - How you installed TVM-Unity (`pip`, source): pip
 - Python version (e.g. 3.10): 3.10
 - GPU driver version (if applicable):
 - CUDA/cuDNN version (if applicable):
 - TVM Unity Hash Tag (`python -c "import tvm; print('\n'.join(f'{k}: {v}' for k, v in tvm.support.libinfo().items()))"`, applicable if you compile models):

> USE_NVTX: OFF
> USE_GTEST: AUTO
> SUMMARIZE: OFF
> USE_IOS_RPC: OFF
> USE_MSC: OFF
> USE_ETHOSU:
> CUDA_VERSION: NOT-FOUND
> USE_LIBBACKTRACE: AUTO
> DLPACK_PATH: 3rdparty/dlpack/include
> USE_TENSORRT_CODEGEN: OFF
> USE_THRUST: OFF
> USE_TARGET_ONNX: OFF
> USE_AOT_EXECUTOR: ON
> BUILD_DUMMY_LIBTVM: OFF
> USE_CUDNN: OFF
> USE_TENSORRT_RUNTIME: OFF
> USE_ARM_COMPUTE_LIB_GRAPH_EXECUTOR: OFF
> USE_CCACHE: AUTO
> USE_ARM_COMPUTE_LIB: OFF
> USE_CPP_RTVM:
> USE_OPENCL_GTEST: /path/to/opencl/gtest
> USE_MKL: OFF
> USE_PT_TVMDSOOP: OFF
> MLIR_VERSION: NOT-FOUND
> USE_CLML: OFF
> USE_STACKVM_RUNTIME: OFF
> USE_GRAPH_EXECUTOR_CUDA_GRAPH: OFF
> ROCM_PATH: /opt/rocm
> USE_DNNL: OFF
> USE_VITIS_AI: OFF
> USE_MLIR: OFF
> USE_RCCL: OFF
> USE_LLVM: llvm-config --link-static
> USE_VERILATOR: OFF
> USE_TF_TVMDSOOP: OFF
> USE_THREADS: ON
> USE_MSVC_MT: OFF
> BACKTRACE_ON_SEGFAULT: OFF
> USE_GRAPH_EXECUTOR: ON
> USE_NCCL: OFF
> USE_ROCBLAS: OFF
> GIT_COMMIT_HASH: 0c5a56a5408215b01c756cbbaa8336421a992a6a
> USE_VULKAN: ON
> USE_RUST_EXT: OFF
> USE_CUTLASS: OFF
> USE_CPP_RPC: OFF
> USE_HEXAGON: OFF
> USE_CUSTOM_LOGGING: OFF
> USE_UMA: OFF
> USE_FALLBACK_STL_MAP: OFF
> USE_SORT: ON
> USE_RTTI: ON
> GIT_COMMIT_TIME: 2023-11-16 19:18:47 -0500
> USE_HEXAGON_SDK: /path/to/sdk
> USE_BLAS: none
> USE_ETHOSN: OFF
> USE_LIBTORCH: OFF
> USE_RANDOM: ON
> USE_CUDA: OFF
> USE_COREML: OFF
> USE_AMX: OFF
> BUILD_STATIC_RUNTIME: OFF
> USE_CMSISNN: OFF
> USE_KHRONOS_SPIRV: OFF
> USE_CLML_GRAPH_EXECUTOR: OFF
> USE_TFLITE: OFF
> USE_HEXAGON_GTEST: /path/to/hexagon/gtest
> PICOJSON_PATH: 3rdparty/picojson
> USE_OPENCL_ENABLE_HOST_PTR: OFF
> INSTALL_DEV: OFF
> USE_PROFILER: ON
> USE_NNPACK: OFF
> LLVM_VERSION: 17.0.5
> USE_OPENCL: OFF
> COMPILER_RT_PATH: 3rdparty/compiler-rt
> RANG_PATH: 3rdparty/rang/include
> USE_SPIRV_KHR_INTEGER_DOT_PRODUCT: OFF
> USE_OPENMP: OFF
> USE_BNNS: OFF
> USE_CUBLAS: OFF
> USE_METAL: OFF
> USE_MICRO_STANDALONE_RUNTIME: OFF
> USE_HEXAGON_EXTERNAL_LIBS: OFF
> USE_ALTERNATIVE_LINKER: AUTO
> USE_BYODT_POSIT: OFF
> USE_HEXAGON_RPC: OFF
> USE_MICRO: OFF
> DMLC_PATH: 3rdparty/dmlc-core/include
> INDEX_DEFAULT_I64: ON
> USE_RELAY_DEBUG: OFF
> USE_RPC: ON
> USE_TENSORFLOW_PATH: none
> TVM_CLML_VERSION:
> USE_MIOPEN: OFF
> USE_ROCM: OFF
> USE_PAPI: OFF
> USE_CURAND: OFF
> TVM_CXX_COMPILER_PATH: C:/Program Files/Microsoft Visual Studio/2022/Enterprise/VC/Tools/MSVC/14.37.32822/bin/HostX64/x64/cl.exe
> HIDE_PRIVATE_SYMBOLS: OFF

 - Any other relevant information:

## Additional context

I tried to parse the convert_weight in `mlc_chat`. I wanted to change the shape of the tensor before `_check_param`, but I couldn't because I wasn't familiar with some tvm commands. I really need to learn some tvm commands, but I don't have time right now.
Thanks

**Comment by Hzfengsy** [comment]
Sorry to hear that, but it's your private model so we cannot reproduce the error. Love to help if you could provide a reproducible model

**Comment by junrushao** [comment]
It’s weird though. Is the model architecture identical to Llama2-13B? If so, I do believe that MLC works pretty smoothly with a 13B Llama, so there’s no reason it just fails on another identical one

**Comment by Tao-begd** [comment]
> It’s weird though. Is the model architecture identical to Llama2-13B? If so, I do believe that MLC works pretty smoothly with a 13B Llama, so there’s no reason it just fails on another identical one

I compared the difference in this parameter between my model and the original model
my model
> model.layers.30.self_attn.q_proj.weight :  torch.Size([5120, 5120])
> model.layers.30.self_attn.k_proj.weight :  torch.Size([5120, 5120])
> model.layers.30.self_attn.v_proj.weight :  torch.Size([5120, 5120])
> model.layers.30.self_attn.o_proj.weight :  torch.Size([5120, 5120])
> model.layers.30.self_attn.rotary_emb.inv_freq :  torch.Size([64])
> model.layers.30.mlp.gate_proj.weight :  torch.Size([13696, 5120])
> model.layers.30.mlp.down_proj.weight :  torch.Size([5120, 13696])
> model.layers.30.mlp.up_proj.weight :  torch.Size([13696, 5120])
> model.layers.30.input_layernorm.weight :  torch.Size([5120])
> model.layers.30.post_attention_layernorm.weight :  torch.Size([5120])

original model

> model.layers.30.self_attn.q_proj.weight :  torch.Size([5120, 5120])
> model.layers.30.self_attn.k_proj.weight :  torch.Size([5120, 5120])
> model.layers.30.self_attn.v_proj.weight :  torch.Size([5120, 5120])
> model.layers.30.self_attn.o_proj.weight :  torch.Size([5120, 5120])
> model.layers.30.self_attn.rotary_emb.inv_freq :  torch.Size([64])
> model.layers.30.mlp.gate_proj.weight :  torch.Size([13824, 5120])
> model.layers.30.mlp.up_proj.weight :  torch.Size([13824, 5120])
> model.layers.30.mlp.down_proj.weight :  torch.Size([5120, 13824])
> model.layers.30.input_layernorm.weight :  torch.Size([5120])
> model.layers.30.post_attention_layernorm.weight :  torch.Size([5120])

The tensor size of `model.layers.30.self_attn.q_proj.weight `  is the same.
But it looks like the `model.layers.30.mlp`  is different, does it need to be aligned with the original model?
Do you have any good suggestions for me to modify the shape of this tensor if it needs to be aligned?
@junrushao

**Comment by junrushao** [comment]
If the model architecture differs, you will have to change model definition in https://github.com/mlc-ai/mlc-llm/blob/main/python/mlc_chat/compiler/model/llama/llama_model.py

**Comment by Tao-begd** [comment]
> If the model architecture differs, you will have to change model definition in https://github.com/mlc-ai/mlc-llm/blob/main/python/mlc_chat/compiler/model/llama/llama_model.py

Our model adopts MHA, while the open source llama2 seems to use GQA. I do not know whether this is the reason for the compilation failure.
Sorry, I tried to modify `llama_model.py`, but obviously I can't at the moment

**Comment by Tao-begd** [comment]
I sorted out https://github.com/mlc-ai/mlc-llm/blob/main/python/mlc_chat/compiler/model/llama/llama_model.py in detail.
I see where the problem is. In `LlamaFFN`, `intermediate_size` is defined as the output layer. However, it is computed using `intermediate_size` and `tensor_parallel_shards`. And I gave `intermediate_size` wrong in the configuration file.
The model structure given above actually contains the value of this parameter, the original llama2 is 13824 and my model is 13696.
I fixed the error in `config.json` and the compilation was successful.
Thanks for your help @junrushao
